New Indonesian Translation: Sharing State Between Components (#370)
reactjs/id.react.dev: src/content/learn/sharing-state-between-components.md

---
title: Berbagi State Antar Komponen
---

<Intro>

Terkadang, Anda ingin 2 komponen selalu berubah secara bersamaan. Untuk melakukannya, hapus *state* dari kedua komponen, pindahkan ke komponen induk terdekat, dan kemudian oper ke komponen tersebut melalui *props*. Ini dikenal sebagai *mengangkat state ke atas*, dan ini adalah salah satu hal yang paling umum yang akan Anda lakukan saat menulis kode React.

</Intro>

<YouWillLearn>

> - Bagaimana cara berbagi state antar komponen dengan *lifting state up*

Reviewer: ```suggestion
- Bagaimana cara berbagi *state* antar komponen dengan menaikkan *state* ke atas (*lifting state up*)
```

---

> title: Berbagi State Antar Komponen
> 
> ---

Reviewer 2: ```suggestion
title: Berbagi State Antar Komponen
---
```

Make sure to not mess around with the other Markdown syntax.

<Intro>

Terkadang, Anda ingin 2 komponen selalu berubah secara bersamaan. Untuk melakukannya, hapus *state* dari kedua komponen, pindahkan ke komponen induk terdekat, dan kemudian oper ke komponen tersebut melalui *props*. Ini dikenal sebagai mengangkat *state* keatas, dan ini adalah salah satu hal yang paling umum yang akan Anda lakukan saat menulis kode React.

</Intro>

<YouWillLearn>

- Bagaimana cara berbagi state antar komponen dengan *lifting state up*

> - Apa itu komponen terkendali dan tidak terkendali

Reviewer 2: According to the [Glossary](https://github.com/reactjs/id.react.dev/wiki/Glosarium) the term `controlled component` shouldn't be translated, but that was in the past. Should we change how we translate this? @mhaidarhanif @r17x @zainfathoni

Reviewer 3: Yes, I think we should translate it. "Komponen terkendali" is already explicit enough and easily understood.

Reviewer 3: I would prefer "tak terkendali" to "tidak terkendali", though. It's more concise and simpler.

Reviewer: @resir014 @zainfathoni Okay how about:

- `komponen terkendali (*controlled component*)`
- `komponen tak erkendali (*uncontrolled component*)`

As usual, the English words are introduced once in the beginning. Then after that just use the translated version.

Author: it is better with `tak terkendali` mas

<Intro>

> Terkadang, Anda ingin 2 komponen selalu berubah secara bersamaan. Untuk melakukannya, hapus *state* dari kedua komponen, pindahkan ke komponen induk terdekat, dan kemudian oper ke komponen tersebut melalui *props*. Ini dikenal sebagai mengangkat *state* keatas, dan ini adalah salah satu hal yang paling umum yang akan Anda lakukan saat menulis kode React.

Reviewer: Indonesian `ke` need to be separated.

```suggestion
Terkadang, Anda ingin 2 komponen selalu berubah secara bersamaan. Untuk melakukannya, hapus *state* dari kedua komponen, pindahkan ke komponen induk terdekat, dan kemudian oper ke komponen tersebut melalui *props*. Ini dikenal sebagai mengangkat *state* ke atas, dan ini adalah salah satu hal yang paling umum yang akan Anda lakukan saat menulis kode React.
```

  return (
    <>
      <h2>Almaty, Kazakhstan</h2>
      <Panel title="Tentang">
        Dengan populasi sekitar 2 juta, Almaty adalah kota terbesar di Kazakhstan. Dari tahun 1929 hingga 1997, itu adalah ibu kota negara tersebut.
      </Panel>
      <Panel title="Etimologi">

>         Nama berasal dari <span lang="kk-KZ">алма</span>, kata Kazakh untuk "apel" dan sering diterjemahkan sebagai "penuh dengan apel". Faktanya, wilayah sekitar Almaty dipercaya sebagai rumah leluhur apel, dan <i lang="la">Malus sieversii</i> si liar dianggap sebagai kandidat yang mungkin untuk leluhur apel domestik modern.

Reviewer 2: ```suggestion
        Nama berasal dari <span lang="kk-KZ">алма</span>, kata Kazakh untuk "apel" dan sering diterjemahkan sebagai "penuh dengan apel". Faktanya, wilayah sekitar Almaty dipercaya sebagai rumah leluhur apel, dan tanaman liar <i lang="la">Malus sieversii</i> dianggap sebagai kandidat yang mungkin untuk leluhur apel domestik modern.
```

</Sandpack>

Perhatikan bagaimana menekan tombol satu panel tidak memengaruhi panel lainnya--mereka independen.

<DiagramGroup>

<Diagram name="sharing_state_child" height={367} width={477} alt="Diagram yang menunjukkan pohon dari tiga komponen, satu induk yang diberi label Accordion dan dua anak yang diberi label Panel dan dua komponen anak yang berlabel Panel. Kedua komponen Panel berisi isActive dengan nilai false.">

Awalnya, setiap `Panel` memiliki state `isActive` dengan nilai `false`, sehingga keduanya terlihat tertutup

</Diagram>

<Diagram name="sharing_state_child_clicked" height={367} width={480} alt="Diagram yang sama seperti sebelumnya, dengan isActive dari komponen Panel anak pertama yang disorot menunjukkan klik dengan nilai isActive diatur ke true. Komponen Panel anak kedua masih berisi nilai false." >

Menekan tombol `Panel` mana pun hanya akan memperbarui *state* `isActive` dari `Panel` itu sendiri

</Diagram>

</DiagramGroup>

> **Tetapi sekarang katakanlah Anda ingin mengubah panel tersebut sehingga hanya satu panel yang dibuka pada satu waktu.** Dengan desain diatas, membuka panel kedua harus menutup panel pertama. Bagaimanakah Anda akan melakukannya?

Reviewer 3: ```suggestion
**Tetapi sekarang katakanlah Anda ingin mengubah panel tersebut sehingga hanya satu panel yang dibuka pada satu waktu.** Dengan desain di atas, membuka panel kedua berarti secara otomatis menutup panel pertama. Bagaimanakah Anda akan melakukannya?
```

I think we can paraphrase it better to make it more understandable.

Untuk mengkoordinasikan kedua panel ini, Anda perlu "mengangkat *state* mereka" ke komponen induk dalam tiga langkah:

1. **Hapus** *state* dari komponen anak.

> 2. **Oper** data dari komponen induk.

Reviewer 3: ```suggestion
2. **Oper** data yang dituliskan langsung di dalam kode (*hardcoded*) dari komponen induk.
```

I don't think we should leave the "hardcoded" word out of the translated content because it's part of the vital aspect of this step, and we will eventually replace the hardcoded data with a state in the next step.

3. **Tambahkan** *state* ke komponen induk dan oper bersamaan dengan *Event Handlers*.

Cara ini akan memungkinkan komponen `Accordion` untuk mengkoordinasikan kedua `Panel` dan hanya membuka satu panel pada satu waktu.

### Langkah 1: Hapus state dari komponen anak {/*step-1-remove-state-from-the-child-components*/}

Anda akan memberikan kontrol `isActive` dari `Panel` ke komponen induknya. Ini berarti komponen induk akan mengoper `isActive` ke `Panel` sebagai *prop*. Mulai dengan **menghapus baris ini** dari komponen `Panel`:

```js
const [isActive, setIsActive] = useState(false);
```

Lalu, tambahkan `isActive` ke daftar *prop* `Panel`:

```js
function Panel({ title, children, isActive }) 
```

> Sekarang komponen induk `Panel` dapat *mengontrol* `isActive` dengan [mengoper sebagai prop.](/learn/passing-props-to-a-component) Sebaliknya, komponen `Panel` sekarang tidak memiliki *kontrol* atas nilai `isActive` - sekarang terserah komponen induk!

Reviewer 3: ```suggestion
Sekarang komponen induk `Panel` dapat *mengontrol* `isActive` dengan [mengoper sebagai prop.](/learn/passing-props-to-a-component) Sebaliknya, komponen `Panel` sekarang tidak memiliki *kontrol* atas nilai `isActive`--sekarang terserah komponen induk!
```

> ### Langkah 2: Oper data yang diperlukan dari komponen induk {/*step-2-pass-hardcoded-data-from-the-common-parent*/}

Reviewer 3: ```suggestion
### Langkah 2: Oper data yang dituliskan langsung di dalam kode dari komponen induk {/*step-2-pass-hardcoded-data-from-the-common-parent*/}
```

Let's preserve the "hardcoded" meaning in the translated content because it's part of the vital aspect of this step, and we will eventually replace the hardcoded data with a state in the next step.

Untuk mengangkat *state*, Anda harus menemukan komponen induk yang paling dekat dari *kedua* komponen anak yang ingin Anda koordinasikan:

> * `Accordion` *(Komponen induk terdekat)*

Reviewer 3: ```suggestion
* `Accordion` *(komponen induk terdekat)*
```

Again, it's not capitalized initially, so let's keep it that way.

</Sandpack>

> Perhatikan bagaimana menekan tombol satu panel tidak memengaruhi panel lainnya - mereka independen.

Reviewer 2: ```suggestion
Perhatikan bagaimana menekan tombol satu panel tidak memengaruhi panel lainnya--mereka independen.
```

Pastikan tanda bacanya tidak ada yang berubah.

</Diagram>

</DiagramGroup>

**Tetapi sekarang katakanlah Anda ingin mengubah panel tersebut sehingga hanya satu panel yang dibuka pada satu waktu.** Dengan desain diatas, membuka panel kedua harus menutup panel pertama. Bagaimana Anda melakukannya?

Untuk mengkoordinasikan kedua panel ini, Anda perlu "mengangkat *state* mereka" ke komponen induk dalam tiga langkah:

1. **Hapus** *state* dari komponen anak.
2. **Oper** data dari komponen induk.
3. **Tambahkan** *state* ke komponen induk dan oper bersamaan dengan *Event Handlers*.

Cara ini akan memungkinkan komponen `Accordion` untuk mengkoordinasikan kedua `Panel` dan hanya membuka satu panel pada satu waktu.

### Langkah 1: Hapus state dari komponen anak {/*step-1-remove-state-from-the-child-components*/}

Anda akan memberikan kontrol `isActive` dari `Panel` ke komponen induknya. Ini berarti komponen induk akan mengoper `isActive` ke `Panel` sebagai *prop*. Mulai dengan **menghapus baris ini** dari komponen `Panel`:

```js
const [isActive, setIsActive] = useState(false);
```

Lalu, tambahkan `isActive` ke daftar *prop* `Panel`:

```js
function Panel({ title, children, isActive }) 
```

> Sekarang komponen induk `Panel` dapat *mengontrol* `isActive` dengan [mengoper sebagai prop.](/learn/passing-props-to-a-component) Sebaliknya, komponen `Panel` sekarang tidak memiliki *kontrol* atas nilai `isActive` - sekarang terserah komponen induk!
> 

Reviewer 2: ```suggestion
Sekarang komponen induk `Panel` dapat *mengontrol* `isActive` dengan [mengoper sebagai prop.](/learn/passing-props-to-a-component) Sebaliknya, komponen `Panel` sekarang tidak memiliki *kontrol* atas nilai `isActive` - sekarang terserah komponen induk!
```

Extra line here.

</Sandpack>

> Coba edit nilai `isActive` yang telah ditentukan sebelumnya di komponen `Accordion` dan lihat hasilnya di layar.

Reviewer 3: ```suggestion
Coba ubah nilai `isActive` yang dituliskan langsung di dalam kode komponen `Accordion` dan lihat hasilnya di layar.
```

Let's preserve the "hardcoded" meaning in the translated content because it's part of the vital aspect of this step, and we will eventually replace the hardcoded data with a state in the next step.

### Langkah 3: Tambahkan state ke komponen induk {/*step-3-add-state-to-the-common-parent*/}

> Memindahkan *state* keatas sering mengubah sifat apa yang Anda simpan sebagai *state*. 

Reviewer 3: ```suggestion
Memindahkan *state* ke atas seringkali mengubah sifat dari apa yang Anda simpan sebagai *state*. 
```

> Dalam contoh ini, Anda ingin mengubah `Accordion` sehingga hanya satu panel yang dapat dibuka pada satu waktu. Ini berarti bahwa komponen induk `Accordion` perlu melacak *panel mana* yang sedang aktif. Alih-alih nilai `boolean`, ini dapat menggunakan angka sebagai indeks `Panel` aktif untuk variabel *state*:

Reviewer 3: ```suggestion
Dalam contoh ini, Anda ingin mengubah `Accordion` sehingga hanya satu panel yang dapat dibuka pada satu waktu. Ini berarti bahwa komponen induk `Accordion` perlu melacak *panel mana* yang sedang aktif. Alih-alih nilai `boolean`, ia dapat menggunakan angka sebagai indeks `Panel` aktif untuk variabel *state*:
```

Because "it" refers to the parent component, let's personify it.

```js
const [activeIndex, setActiveIndex] = useState(0);
```

Ketika `activeIndex` bernilai `0`, panel pertama aktif, dan ketika bernilai `1`, panel kedua aktif.

> Menekan tombol "Tampilkan" di salah satu `Panel` perlu mengubah indeks aktif di `Accordion`. Sebuah `Panel` tidak dapat mengatur *state* `activeIndex` secara langsung karena didefinisikan di dalam `Accordion`. Komponen `Accordion` perlu *mengeksplisitkan* komponen `Panel` untuk mengubah *state*-nya dengan [mengoper event handler sebagai prop](/learn/responding-to-events#passing-event-handlers-as-props):

Reviewer 3: ```suggestion
Menekan tombol "Tampilkan" di salah satu `Panel` perlu mengubah indeks aktif di `Accordion`. Sebuah `Panel` tidak dapat mengatur *state* `activeIndex` secara langsung karena ia didefinisikan di dalam `Accordion`. Komponen `Accordion` perlu *secara ekplisit mengizinkan* komponen `Panel` untuk mengubah *state*-nya dengan [mengoper *event handler* sebagai prop](/learn/responding-to-events#passing-event-handlers-as-props):
```

- Because "it" refers to the state, let's personify it.
- "explicitly allow" should translate to "*secara eksplisit mengizinkan*".
- make "event handler" italic.

    <>
      <h2>Almaty, Kazakhstan</h2>
      <Panel
        title="Tentang"
        isActive={activeIndex === 0}
        onShow={() => setActiveIndex(0)}
      >
        Dengan populasi sekitar 2 juta, Almaty adalah kota terbesar di Kazakhstan. Dari tahun 1929 hingga 1997, itu adalah ibu kota negara tersebut.
      </Panel>
      <Panel
        title="Etimologi"
        isActive={activeIndex === 1}
        onShow={() => setActiveIndex(1)}
      >

>        Nama berasal dari <span lang="kk-KZ">алма</span>, kata Kazakh untuk "apel" dan sering diterjemahkan sebagai "penuh dengan apel". Faktanya, wilayah sekitar Almaty dipercaya sebagai rumah leluhur apel, dan <i lang="la">Malus sieversii</i> si liar dianggap sebagai kandidat yang mungkin untuk leluhur apel domestik modern.

Reviewer 2: ```suggestion
       Nama berasal dari <span lang="kk-KZ">алма</span>, kata Kazakh untuk "apel" dan sering diterjemahkan sebagai "penuh dengan apel". Faktanya, wilayah sekitar Almaty dipercaya sebagai rumah leluhur apel, dan tanaman liar <i lang="la">Malus sieversii</i> dianggap sebagai kandidat yang mungkin untuk leluhur apel domestik modern.
```

</Sandpack>

> Dengan ini terselesaikan pengangkatan *state* ke atas! Memindahkan *state* ke komponen induk memungkinkan Anda untuk mengkoordinasikan kedua panel. Menggunakan indeks aktif daripada dua *flag* "ditampilkan" memastikan bahwa hanya satu panel yang aktif pada satu waktu. Dan mengoper *event handler* ke komponen anak memungkinkan komponen anak untuk mengubah *state* induknya.

Reviewer 3: ```suggestion
Dengan ini, terselesaikanlah pengangkatan *state* ke atas! Memindahkan *state* ke komponen induk memungkinkan Anda untuk mengkoordinasikan kedua panel. Menggunakan indeks aktif alih-alih dua *flag* "ditampilkan" memastikan bahwa hanya satu panel yang aktif pada satu waktu. Dan mengoper *event handler* ke komponen anak memungkinkan komponen anak untuk mengubah *state* induknya.
```

I think it's more accurate to translate "instead of" to "*alih-alih*" rather than *daripada*.

<DiagramGroup>

<Diagram name="sharing_state_parent" height={385} width={487} alt="Diagram yang menunjukkan pohon tiga komponen, satu induk yang diberi label Accordion dan dua anak yang diberi label Panel. Accordion berisi nilai activeIndex nol yang berubah menjadi nilai isActive true yang dilewatkan ke Panel pertama, dan nilai isActive false yang dioper ke Panel kedua." >

Awalnya, nilai `activeIndex` dari `Accordion` adalah `0`, sehingga `Panel` pertama menerima `isActive = true`

</Diagram>

<Diagram name="sharing_state_parent_clicked" height={385} width={521} alt="Diagram yang sama dengan sebelumnya, dengan nilai activeIndex dari komponen induk Accordion yang diberi sorotan menunjukkan klik dengan nilai yang berubah menjadi satu. Aliran ke kedua komponen Panel anak juga disorot, dan nilai isActive yang dioper ke setiap anak diatur ke yang berlawanan: false untuk Panel pertama dan true untuk yang kedua." >

Ketika nilai `activeIndex` dari `Accordion` berubah menjadi `1`, `Panel` kedua menerima `isActive = true` sebagai gantinya

</Diagram>

</DiagramGroup>

<DeepDive>

#### Komponen terkendali dan tak terkendali {/*controlled-and-uncontrolled-components*/}

> Secara umum Anda dapat mengatakan bahwa komponen yang memiliki *state* lokal adalah "tak terkendali". Misalnya, komponen `Panel` asli dengan variabel *state* `isActive` adalah tak terkendali karena induknya tidak dapat mempengaruhi apakah panel aktif atau tidak.

Reviewer 3: ```suggestion
Secara umum Anda dapat menyebut sebuah komponen yang memiliki *state* lokal sebagai "tak terkendali". Misalnya, komponen `Panel` yang asli dengan variabel *state* `isActive` adalah tak terkendali karena induknya tidak dapat mempengaruhi apakah panel aktif atau tidak.
```

> Anda dapat mengatakan bahwa komponen "terkendali" ketika informasi penting di dalamnya didorong oleh *prop* daripada *state* lokalnya sendiri. Ini memungkinkan komponen induk untuk sepenuhnya menentukan perilakunya. Komponen `Panel` akhir dengan *prop* `isActive` dikendalikan oleh komponen `Accordion`.

Reviewer 3: ```suggestion
Sebaliknya, Anda dapat menyebut sebuah komponen "terkendali" ketika informasi penting di dalamnya dikendalikan oleh *prop* daripada *state* lokalnya sendiri. Ini memungkinkan komponen induk untuk sepenuhnya menentukan perilakunya. Komponen `Panel` akhir dengan *prop* `isActive` dikendalikan oleh komponen `Accordion`.
```

> Komponen tak terkendali lebih mudah digunakan dalam induknya karena membutuhkan konfigurasi yang lebih sedikit. Tetapi kurang fleksibel ketika Anda ingin mengkoordinasikan mereka bersama. Komponen terkendali sepenuhnya fleksibel, tetapi mereka membutuhkan komponen induk untuk sepenuhnya mengkonfigurasi mereka dengan *props*.

Reviewer 3: ```suggestion
Komponen-komponen tak terkendali lebih mudah digunakan dalam induknya karena membutuhkan konfigurasi yang lebih sedikit. Tetapi mereka kurang fleksibel ketika Anda ingin mengkoordinasikannya bersamaan. Komponen-komponen terkendali sepenuhnya fleksibel, tetapi mereka membutuhkan komponen induk untuk sepenuhnya mengkonfigurasi mereka dengan *props*.
```

Because "they" refers to the components, let's personify it.

> Pada praktiknya, "terkendali" dan "tak terkendali" bukan istilah teknis yang ketat--setiap komponen biasanya memiliki beberapa campuran dari kedua *state* lokal dan *props*. Namun, ini adalah cara yang berguna untuk berbicara tentang bagaimana komponen dirancang dan kemampuan apa yang mereka tawarkan.

Reviewer 3: ```suggestion
Pada praktiknya, "terkendali" dan "tak terkendali" bukanlah istilah teknis yang ketat--setiap komponen biasanya memiliki beberapa campuran dari kedua *state* lokal dan *props*. Namun, ini adalah cara yang berguna untuk berbicara tentang bagaimana komponen dirancang dan kemampuan apa yang mereka tawarkan.
```

Adding a "*-lah" particle here emphasizes the meaning of the sentence.

> Ketika menulis komponen, pertimbangkan informasi apa di dalamnya yang harus dikendalikan (melalui *props*), dan informasi apa yang harus tak terkendali (melalui *state*). Tetapi Anda selalu dapat mengubah pikiran Anda dan melakukan refaktorasi nanti.

Reviewer 3: ```suggestion
Ketika menulis komponen, pertimbangkan informasi manakah di dalamnya yang seharusnya dikendalikan (melalui *props*), dan informasi manakah yang seharusnya tak terkendali (melalui *state*). Tetapi Anda selalu dapat mengubah pikiran Anda dan melakukan refaktorasi belakangan.
```

I think "which" translates to "*manakah*" better than "*apa*". And "should" translates to "*seharusnya*" rather than "*harus*", because "*harus*" is the literal translation for "must".

</DeepDive>

## Satu sumber kebenaran untuk setiap state {/*a-single-source-of-truth-for-each-state*/}

> Pada aplikasi React, banyak komponen akan memiliki *state* mereka sendiri. Beberapa *state* mungkin "hidup" dekat dengan komponen daun (komponen di bagian bawah pohon) seperti masukan. *State* lainnya mungkin "hidup" lebih dekat ke bagian atas aplikasi. Misalnya, bahkan pustaka rute pada sisi klien biasanya diimplementasikan dengan menyimpan rute saat ini di *state* React, dan mengopernya dengan *props*!

Reviewer 3: ```suggestion
Pada aplikasi React, banyak komponen akan memiliki *state* mereka sendiri. Beberapa *state* mungkin "hidup" dekat dengan komponen daun (komponen di bagian bawah pohon) seperti masukan-masukan (*inputs*). *State* lainnya mungkin "hidup" lebih dekat ke bagian atas aplikasi. Misalnya, bahkan pustaka-pustaka rute pada sisi klien (*client-side routing libraries*) biasanya diimplementasikan dengan menyimpan rute saat ini di *state* React, dan mengopernya dengan *props*!
```

I think those words need its original English phrases to explain their meaning better.

**Untuk setiap potongan *state* yang unik, Anda akan memilih komponen yang "memilikinya".** Prinsip ini juga dikenal sebagai memiliki ["sumber kebenaran tunggal".](https://en.wikipedia.org/wiki/Single_source_of_truth) Ini tidak berarti bahwa semua *state* berada di satu tempat--tetapi bahwa untuk _setiap_ potongan *state*, ada _komponen_ tertentu yang memegang potongan informasi itu. Alih-alih menduplikasi *state* yang sama diantara komponen, *angkatlah* *state* tersebut ke induk mereka, dan oper ke anak-anak yang membutuhkannya.

> Aplikasi Anda akan berubah saat Anda bekerja di dalamnya. Biasanya Anda akan memindahkan *state* ke bawah atau kembali ke atas saat Anda masih mencari tahu di mana setiap potongan *state* "hidup". Semua ini adalah bagian dari proses!

Reviewer 3: ```suggestion
Aplikasi Anda akan berubah saat Anda mengerjakannya. Biasanya Anda akan memindahkan *state* ke bawah atau kembali ke atas saat Anda masih mencari tahu di mana setiap potongan *state* "hidup". Semua ini adalah bagian dari proses!
```
